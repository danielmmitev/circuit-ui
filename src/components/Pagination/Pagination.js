import React from 'react';
import PropTypes from 'prop-types';
import * as PaginationService from './PaginationService';
import PaginationContainer from './PaginationContainer';
import PaginationButton from './PaginationButton';

/**
 * Pagination is a component to show pages numbers calculate dinamically.
 */

const Pagination = ({
  page,
  perPage,
  total,
  onChange,
  nextLabel,
  previousLabel
}) => {
  const PageButton = ({ value }) => {
    const plain = page !== value;
    return <PaginationButton plain={plain} value={value} onClick={onChange} />;
  };

  PageButton.propTypes = {
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired
  };

  const totalPages = PaginationService.calculatePages(total, perPage);

  if (totalPages < 2) {
    return null;
  }

  if (totalPages <= 5) {
    return (
      <PaginationContainer
        page={page}
        totalPages={totalPages}
        onChange={onChange}
        nextLabel={nextLabel}
        previousLabel={previousLabel}
      >
        {PaginationService.createEmptyArrayFromNumber(totalPages).map(
          (item, index) => <PageButton key={item} value={index + 1} />
        )}
      </PaginationContainer>
    );
  }
  const previousValues = PaginationService.arrayOfPreviousValues(
    page,
    totalPages
  );
  const nextValues = PaginationService.arrayOfNextValues(page, totalPages);

  const shouldHavePreviousDots = PaginationService.shouldHavePreviousDots(
    previousValues
  );
  const shouldHaveNextDots = PaginationService.shouldHaveNextDots(
    nextValues,
    totalPages
  );

  const isNotFirstOrLastPage = page !== totalPages && page !== 1;

  return (
    <PaginationContainer
      page={page}
      totalPages={totalPages}
      onChange={onChange}
      nextLabel={nextLabel}
      previousLabel={previousLabel}
    >
      <PageButton value={1} />
      {shouldHavePreviousDots && '...'}
      {previousValues.map(item => <PageButton key={item} value={item} />)}
      {isNotFirstOrLastPage && <PageButton key={page} value={page} />}
      {nextValues.map(item => <PageButton key={item} value={item} />)}
      {shouldHaveNextDots && '...'}
      <PageButton value={totalPages} />
    </PaginationContainer>
  );
};

Pagination.propTypes = {
  /**
   * Active page of pagination
   */
  page: PropTypes.number,
  /**
   * Number of items per page
   */
  perPage: PropTypes.number,
  /**
   * Total of items inside the pagination
   */
  total: PropTypes.number.isRequired,
  /**
   * Callback for when an page is changed
   */
  onChange: PropTypes.func.isRequired,
  /**
   * Label to be used on buttomn of next
   */
  nextLabel: PropTypes.string,
  /**
   * Label to be used on buttomn of previous
   */
  previousLabel: PropTypes.string
};

Pagination.defaultProps = {
  page: 1,
  perPage: 50,
  nextLabel: 'Next',
  previousLabel: 'Previous'
};

/**
 * @component Pagination
 */
export default Pagination;
