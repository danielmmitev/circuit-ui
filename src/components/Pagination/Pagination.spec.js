import React from 'react';
import { identity } from 'lodash/fp';
import Pagination from './Pagination';

describe('Pagination', () => {
  describe('styles', () => {
    describe('With 5 or less pages', () => {
      const component = shallow(
        <Pagination
          page={1}
          total={250}
          translate={identity}
          onChange={identity}
        />
      );
      expect(component).toMatchSnapshot();
    });

    describe('With more than 5 pages', () => {
      describe('when user is on page one', () => {
        const component = shallow(
          <Pagination
            page={1}
            total={500}
            translate={identity}
            onChange={identity}
          />
        );
        expect(component).toMatchSnapshot();
      });

      describe('when user is on page 2', () => {
        const component = shallow(
          <Pagination
            page={2}
            total={500}
            translate={identity}
            onChange={identity}
          />
        );
        expect(component).toMatchSnapshot();
      });

      describe('when user is on last page', () => {
        const component = shallow(
          <Pagination
            page={10}
            total={500}
            translate={identity}
            onChange={identity}
          />
        );
        expect(component).toMatchSnapshot();
      });

      describe('when user is last but one page', () => {
        const component = shallow(
          <Pagination
            page={9}
            total={500}
            translate={identity}
            onChange={identity}
          />
        );
        expect(component).toMatchSnapshot();
      });

      describe('when user is on page in the middle', () => {
        const component = shallow(
          <Pagination
            page={5}
            total={500}
            translate={identity}
            onChange={identity}
          />
        );
        expect(component).toMatchSnapshot();
      });
    });
  });

  describe('accessibility', () => {
    it('should meet accessibility guidelines', async () => {
      const wrapper = renderToHtml(
        <Pagination
          page={1}
          total={250}
          translate={identity}
          onChange={identity}
        />
      );
      const actual = await axe(wrapper);
      expect(actual).toHaveNoViolations();
    });
  });
});
