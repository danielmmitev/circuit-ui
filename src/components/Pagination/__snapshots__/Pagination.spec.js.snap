// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<PaginationContainer
  nextLabel="Next"
  onChange={[Function]}
  page={1}
  previousLabel="Previous"
  totalPages={5}
>
  <PageButton
    key="0"
    value={1}
  />
  <PageButton
    key="1"
    value={2}
  />
  <PageButton
    key="2"
    value={3}
  />
  <PageButton
    key="3"
    value={4}
  />
  <PageButton
    key="4"
    value={5}
  />
</PaginationContainer>
`;

exports[` 2`] = `
<PaginationContainer
  nextLabel="Next"
  onChange={[Function]}
  page={1}
  previousLabel="Previous"
  totalPages={10}
>
  <PageButton
    value={1}
  />
  <PageButton
    key="2"
    value={2}
  />
  <PageButton
    key="3"
    value={3}
  />
  <PageButton
    key="4"
    value={4}
  />
  ...
  <PageButton
    value={10}
  />
</PaginationContainer>
`;

exports[` 3`] = `
<PaginationContainer
  nextLabel="Next"
  onChange={[Function]}
  page={2}
  previousLabel="Previous"
  totalPages={10}
>
  <PageButton
    value={1}
  />
  <PageButton
    key="2"
    value={2}
  />
  <PageButton
    key="3"
    value={3}
  />
  <PageButton
    key="4"
    value={4}
  />
  ...
  <PageButton
    value={10}
  />
</PaginationContainer>
`;

exports[` 4`] = `
<PaginationContainer
  nextLabel="Next"
  onChange={[Function]}
  page={10}
  previousLabel="Previous"
  totalPages={10}
>
  <PageButton
    value={1}
  />
  ...
  <PageButton
    key="7"
    value={7}
  />
  <PageButton
    key="8"
    value={8}
  />
  <PageButton
    key="9"
    value={9}
  />
  <PageButton
    value={10}
  />
</PaginationContainer>
`;

exports[` 5`] = `
<PaginationContainer
  nextLabel="Next"
  onChange={[Function]}
  page={9}
  previousLabel="Previous"
  totalPages={10}
>
  <PageButton
    value={1}
  />
  ...
  <PageButton
    key="7"
    value={7}
  />
  <PageButton
    key="8"
    value={8}
  />
  <PageButton
    key="9"
    value={9}
  />
  <PageButton
    value={10}
  />
</PaginationContainer>
`;

exports[` 6`] = `
<PaginationContainer
  nextLabel="Next"
  onChange={[Function]}
  page={5}
  previousLabel="Previous"
  totalPages={10}
>
  <PageButton
    value={1}
  />
  ...
  <PageButton
    key="4"
    value={4}
  />
  <PageButton
    key="5"
    value={5}
  />
  <PageButton
    key="6"
    value={6}
  />
  ...
  <PageButton
    value={10}
  />
</PaginationContainer>
`;
